### PR TITLE
[not for review] parallel_device_ids demo

### PR DESCRIPTION
**Pull request by DuYicong515**
## What does this PR do?

Fixes #\<issue_number>

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/callbacks/gpu_stats_monitor.py [hunk 1]
```diff
@@ -133,8 +134,16 @@ def setup(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule", stage: O
             )
 
         # The logical device IDs for selected devices
-        # ignoring mypy check because `trainer.data_parallel_device_ids` is None when using CPU
-        self._device_ids = sorted(set(trainer.data_parallel_device_ids))  # type: ignore
+        # this is going to be wrong since parallel_device_ids=[0] if we use GPU + SingleDeviceStrategy
+        # However, parallel_devices now only exists in ParallelStrategy
+        # self._device_ids changes from [0] to [] (L121)
+        if hasattr(trainer.strategy, "parallel_devices"):
```
Comment (DuYicong515) [inline pytorch_lightning/callbacks/gpu_stats_monitor.py:140]: @four4fish So this migration is apparently wrong with  GPU + SingleDeviceStrategy, self._device_ids changes from [0] to []


Currently parallel_devices is only with ParallelStrategy https://github.com/PyTorchLightning/pytorch-lightning/blob/550d3a640d1bfeb731a16b2996d3160f0f7eb071/pytorch_lightning/strategies/parallel.py#L75-L77

Should I upstream it to Strategy and implement it?

Comment (four4fish) [reply pytorch_lightning/callbacks/gpu_stats_monitor.py:140]: This file is deprecated, will be removed right after 1.6 release. We only need to keep logic as-is.
> Should I upstream it to Strategy and implement it?

no, we shouldn't

Comment (four4fish) [reply pytorch_lightning/callbacks/gpu_stats_monitor.py:140]: Here you can still call trainer.data_parallel_device_ids
This file will be removed before trainer.data_parallel_device_ids removes

Comment (DuYicong515) [reply pytorch_lightning/callbacks/gpu_stats_monitor.py:140]: okay, so I'll keep this. 

However, I feel `getattr(trainer.strategy, "parallel_devices", [])` behaviour isn't the same as `trainer._accelerator_connector.parallel_devices` on GPU + SingleDeviceStrategy. Thus `trainer._accelerator_connector.parallel_device_ids` behaviour will change as well, if we rely on `trainer.strategy.parallel_devices` 

  Do we still want to make the change?

Comment (DuYicong515) [reply pytorch_lightning/callbacks/gpu_stats_monitor.py:140]: There's still one other place (not test) using trainer._accelerator_connector.parallel_device_ids if we keep this one unchanged.
https://github.com/PyTorchLightning/pytorch-lightning/blob/4c4b9d540f065c41ad288a78110cecbf45b96409/pytorch_lightning/trainer/connectors/logger_connector/logger_connector.py#L226

Comment (four4fish) [reply pytorch_lightning/callbacks/gpu_stats_monitor.py:140]: The goal is to 
1. remove accelerator.parallel_device_ids
2. update trainer.data_parallel_device_ids to reproduce the same logic + add deprecation message
3. update tests with pytest.deprecated...

This file shouldn't change

Comment (four4fish) [reply pytorch_lightning/callbacks/gpu_stats_monitor.py:140]: Oh I see! The logger connector is using parallel_device_ids. 
The [] and None make no different there I think, could you update it to trainer.strategy.parallel_devices to give it a try?

Comment (DuYicong515) [reply pytorch_lightning/callbacks/gpu_stats_monitor.py:140]: If it's using `SingleDeviceStrategy` but with GPU=1 (like in my test case below) it won't have a `trainer.strategy.parallel_devices` since `parallel_devices` is only defined on `ParallelStrategy` so `SingleDeviceStrategy` does not have it.

trainer.strategy.parallel_devices will break on `SingleDeviceStrategy`, however the original `self.trainer._accelerator_connector.parallel_device_ids` returned [0], instead of []/None/Exceptions...

Maybe checkout my test case below?

Comment (DuYicong515) [reply pytorch_lightning/callbacks/gpu_stats_monitor.py:140]: with  trainer = Trainer(gpus=1)
`getattr(trainer.strategy, "parallel_devices", [])` or `getattr(trainer.strategy, "parallel_devices", None)` or `trainer._accelerator_connector.parallel_device_ids`

Comment (four4fish) [reply pytorch_lightning/callbacks/gpu_stats_monitor.py:140]: Just saw this method will be removed after 1.6 too

To handle single device you can do it in step 2 :update trainer.data_parallel_device_ids to reproduce the same logic + add deprecation message

do something similar to https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/trainer/connectors/accelerator_connector.py#L793-L799

The caller side can remain the same

Comment (DuYicong515) [reply pytorch_lightning/callbacks/gpu_stats_monitor.py:140]: Oh I see, I'll handle the case in the new logic then. I thought `trainer.strategy.parallel_devices` should logically be an identical replacement for `trainer._accelerator_connector.parallel_devices`, and I was wrong. 

Thanks for the discussions and insights! @four4fish  I'm clear on this now.
